### PR TITLE
[Microsoft Outlook Calendar] add outlook calendar connector with documentation, test cases & samples 

### PR DESCRIPTION
**Pull request by SkNuwanTissera**
## Purpose
Implementation to support major functionalities of Outlook Calendar.
Test cases to test each function.
Display annotations for enhanced visualization of function names.
Samples to provide easy reference for other developers using the package.
Developer documentation that provides an overview of the connector.

fixes https://github.com/wso2-enterprise/choreo/issues/6908
fixes https://github.com/wso2-enterprise/choreo/issues/6909

## Goals
Implement the Ballerina Outlook Calendar connector to provide easy integration to OneDrive using the Ballerina language.

## Approach
Implement the connector on top of Microsoft Graph is a REST web API that empowers you to access Microsoft Cloud
service resources.
Provide a complete overview of the functionalities of the connector by creating developer documentation.
Add display annotations to all the remote operations and clients.
Implement samples to cover all functionalities.

## Automation tests
 - Unit tests 
   > Done
 - Integration tests
   > Done

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
https://github.com/ballerina-platform/module-ballerinax-microsoft.outlook.calendar/tree/master/samples
 
## Learning
The connector was designed mainly referring to the official documentation of Microsoft Graph is a REST web API v1.0
Office 365 Graph API - https://docs.microsoft.com/en-us/graph/overview
Graph explorer - https://developer.microsoft.com/en-us/graph/graph-explorer
API reference for OneDrive - https://docs.microsoft.com/en-us/graph/onedrive-concept-overview
https://docs.microsoft.com/en-us/onedrive/developer/?view=odsp-graph-online

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+                  REFRESH_TOKEN: ${{ secrets.REFRESH_TOKEN }}
+                  CLIENT_ID: ${{ secrets.CLIENT_ID }}
+                  CLIENT_SECRET: ${{ secrets.CLIENT_SECRET }}
+            - name: Ballerina Push
+              uses: ballerina-platform/ballerina-action/@master
+              with:
+                  args:
+                      push
+              env:
+                  BALLERINA_CENTRAL_ACCESS_TOKEN: ${{ secrets.BALLERINA_CENTRAL_ACCESS_TOKEN }}
```
Comment (Chuhaa) [inline .github/workflows/release.yml:30]:
```suggestion
              env:
                  WORKING_DIR: ./calendar
                  BALLERINA_CENTRAL_ACCESS_TOKEN: ${{ secrets.BALLERINA_CENTRAL_ACCESS_TOKEN }}
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,51 @@
...
+[![GitHub Last Commit](https://img.shields.io/github/last-commit/ballerina-platform/module-ballerinax-microsoft.outlook.calendar.svg)](https://github.com/ballerina-platform/module-ballerinax-microsoft.outlook.calendar/commits/master)
+[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
+
+The `microsoft.outlook.calendar` is a [Ballerina](https://ballerina.io/) connector for Microsoft Outlook Calendar.
+
+[Microsoft Outlook Calendar](https://www.microsoft.com/en-ww/microsoft-365/outlook.calendar) is composed of the calendar and scheduling component of Outlook developed by Microsoft.
+
+This connector provides operations for connecting and interacting with Microsoft Graph API Outlook Calendar endpoints over the network. 
+For more information about configuration and operations, go to the module. 
+- [`microsoft.outlook.calendar`](https://docs.central.ballerina.io/ballerinax/microsoft.outlook.calendar/latest).
```
Comment (Chuhaa) [inline README.md:14]:
```suggestion
- [`microsoft.outlook.calendar`](https://docs.central.ballerina.io/ballerinax/microsoft.outlook.calendar/1.0.1).
```
Mention the exact connector version.

File: calendar/Package.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+#### Compatibility
+|                               | Version               |
+|-------------------------------|-----------------------|
+| Ballerina Language Version    | **Swan Lake Beta 1**  |
+| API Version                   | **v1.0**              |
+
+
+### Report Issues
+To report bugs, request new features, start new discussions, view project boards, etc., go to the [Ballerina connector repository](link)
```
Comment (Chuhaa) [inline calendar/Package.md:15]:
```suggestion
To report bugs, request new features, start new discussions, view project boards, etc., go to the [Ballerina Microsoft calendar connector repository](https://github.com/ballerina-platform/module-ballerinax-microsoft.outlook.calendar)
```

File: calendar/README.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
+Ballerina Connector For Microsoft Calendar
```
Comment (Chuhaa) [inline calendar/README.md:1]: README.md inside calendar directory is not required .

File: calendar/client.bal [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    label: "Microsoft Calendar Client",
+    iconPath: "logo.svg"
+}
+public client class Client {
+    http:Client httpClient;
+    Configuration config;
+
+    # Initializes the Outlook Calendar connector client.
+    #
+    # + config - Configurations required to initialize the `Client` endpoint
```
Comment (Chuhaa) [inline calendar/client.bal:33]:
```suggestion
    # + config - Configurations required to initialize the `Client` endpoint
    # + return - An error on failure of initialization or else `()`
```
Shall we add a comment for return?

File: calendar/client.bal [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+    # Get the properties and relationships of the specified event object.
+    # API doc : https://docs.microsoft.com/en-us/graph/api/event-get
+    #
+    # + eventId - ID of an event. Read-only.  
+    # + timeZone - Preferred Time Zone of the start and end time. (Default : `UTC`)
+    # + contentType - Preferred Content-Type of the body. Values : `text` or `text` (Default : `html`)
+    # + queryParams - Optional query parameters. This method support OData query parameters to customize the response. 
+    # It should be an array of type `string` in the format `<QUERY_PARAMETER_NAME>=<PARAMETER_VALUE>`
+    # **Note:** For more information about query parameters, refer here: 
+    # https://docs.microsoft.com/en-us/graph/query-parameters
```
Comment (Chuhaa) [inline calendar/client.bal:59]: Some API guidelines were provided by doc team. Could you check the https://docs.google.com/document/d/1WZg1ix9MpKDTQ_OOm3NVqhQ5L0VGNaUQ1LRmjp-SMO4/edit?ts=60f6ac03

File: calendar/client.bal [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    @display {label: "List Calendars"}
+    remote isolated function listCalendars(@display {label: "Optional Query Parameters"} string? queryParams = ()) 
+                                            returns stream<Calendar, error>|error {
+        string path = check createUrl([LOGGED_IN_USER, CALENDARS], queryParams);
+        CalendarStream objectInstance = check new (self.config, self.httpClient, path, queryParams);
+        stream<Calendar, error> finalStream = new (objectInstance);
+        return finalStream;
+    }
+}
+
```
Comment (Chuhaa) [inline calendar/client.bal:261]: Remove the extra line

File: calendar/stream_implementer.bal [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+import ballerina/http;
+
+class EventStream {
+    private Event[] currentEntries = [];
+    private string nextLink;
+    private string? timeZone;
+    private string? contentType;
+    private string? queryPraram;
+    Configuration config;
+    int index = 0;
```
Comment (Chuhaa) [inline calendar/stream_implementer.bal:26]: Can't these be private?

File: calendar/tests/test.bal [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/io;
+import ballerina/log;
+import ballerina/test;
+
+configurable string & readonly refreshUrl = ?;
```
Comment (Chuhaa) [inline calendar/tests/test.bal:21]:
```suggestion
configurable string refreshUrl = ?;
```
`string` is inherently immutable. So we don't have to say string & readonly (More info: https://ballerina.io/ballerina-spec/spec.html#readonly-type-descriptor).

File: calendar/utils.bal [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    }
+    json errorPayload = check httpResponse.getJsonPayload();
+    string message = errorPayload.toString(); // Error should be defined as a user defined object
+    return error (message);
+}
+
+isolated function getEventById(http:Client httpClient, string url, string? timeZone, string? contentType) returns Event|error {
+    http:Response response = check httpClient->get(url, preparePreferenceHeaderString(timeZone, contentType));
+    json event = check handleResponse(response);
+    return check event.cloneWithType(Event);
```
Comment (Chuhaa) [inline calendar/utils.bal:94]:
```suggestion
    return check httpClient->get(url, preparePreferenceHeaderString(timeZone, contentType), targetType = Event);
```
Can't we use this?

File: .github/workflows/release.yml (hunk 1 above)
Comment (SkNuwanTissera) [reply .github/workflows/release.yml:30]: Thanks 
Changed as suggested

File: README.md (hunk 2 above)
Comment (SkNuwanTissera) [reply README.md:14]: changed as suggested. /Thanks

File: calendar/Package.md (hunk 3 above)
Comment (SkNuwanTissera) [reply calendar/Package.md:15]: Changed as above. Thanks

File: calendar/README.md (hunk 4 above)
Comment (SkNuwanTissera) [reply calendar/README.md:1]: Removed the file

File: calendar/client.bal (hunk 5 above)
Comment (SkNuwanTissera) [reply calendar/client.bal:33]: Done as above. /Thanks

File: calendar/client.bal (hunk 6 above)
Comment (SkNuwanTissera) [reply calendar/client.bal:60]: Changed the doc comments

File: calendar/client.bal (hunk 7 above)
Comment (SkNuwanTissera) [reply calendar/client.bal:262]: Removed the extra line

File: calendar/stream_implementer.bal (hunk 8 above)
Comment (SkNuwanTissera) [reply calendar/stream_implementer.bal:26]: Marked as private

File: calendar/tests/test.bal (hunk 9 above)
Comment (SkNuwanTissera) [reply calendar/tests/test.bal]: Removed read-only

File: calendar/utils.bal (hunk 10 above)
Comment (SkNuwanTissera) [reply calendar/utils.bal:94]: Resolved in new PR

File: build.gradle [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    publications {
+        mavenJava(MavenPublication) {
+            artifact source: createArtifactZip, extension: 'zip'
+        }
+    }
+
+    repositories {
+        maven {
+            name = "GitHubPackages"
+            url = uri("https://maven.pkg.github.com/ballerina-platform/module-ballerinax-microsoft.calendar")
```
Comment (LakshanSS) [inline build.gradle:87]: Seems the url of the repo is wrong.
Please correct

File: calendar/Package.md [hunk 12]
```diff
@@ -0,0 +1,20 @@
+Connects to Microsoft Outlook Calendar from Ballerina
+
+### Package Overview
+The `outlook.calendar` is a [Ballerina](https://ballerina.io/) connector for Outlook Calendar.
+This package provides the capability to easily access Outlook Calendar.
```
Comment (abeykoon) [inline calendar/Package.md:5]: This introduction should have some value. All come here know the above information.

File: calendar/data_mappings.bal [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// isolated function convertToEvent(map<json> sourceEventObject) returns Event|error {
+//     Event event = check sourceEventObject.cloneWithType(Event);
+//     event.downloadUrl = let var url = sourceEventObject["@microsoft.graph.downloadUrl"] in url is string ?
+//         url : EMPTY_STRING;
+//     return event;
+// }
```
Comment (LakshanSS) [inline calendar/data_mappings.bal:22]: Shall we remove this, if this is not needed ?

File: calendar/stream_implementer.bal [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+        return check self.getAndConvertToCalendarArray(response);
+    }
+
+    isolated function fetchRecordsNext() returns Calendar[]|error {
+        http:Client nextPageClient = check new (self.nextLink, {
+            auth: self.config.clientConfig,
+            secureSocket: self.config?.secureSocketConfig
+        });
+        http:Response response 
+            = check nextPageClient->get(EMPTY_STRING);
```
Comment (LakshanSS) [inline calendar/stream_implementer.bal:145]: Shall we fix the formatting.
Please fix the same issue in other places also

File: calendar/Package.md [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
```
Comment (abeykoon) [inline calendar/Package.md:6]: Somewhere we need to point to module.md file. In overview, we do not provide much detail, but we need to point where this info is. @Chuhaa where should that come?

File: calendar/client.bal [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+    iconPath: "logo.svg"
+}
+public client class Client {
+    http:Client httpClient;
+    Configuration config;
+
+    # Initializes the Outlook Calendar connector client.
+    #
+    # + config - Configurations required to initialize the `Client` endpoint
+    # + return - An error on failure of initialization or else `()`
```
Comment (abeykoon) [inline calendar/client.bal:34]: we do not need `or else `()`` part

File: calendar/client.bal [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+    # API doc : https://docs.microsoft.com/en-us/graph/api/event-get
+    #
+    # + eventId - ID of an event. Read-only.  
+    # + timeZone - Preferred Time Zone of the start and end time. (Default : `UTC`)
+    # + contentType - Preferred Content-Type of the body. Values : `text` or `text` (Default : `html`)
+    # + queryParams - Optional query parameters. This method support OData query parameters to customize the response. 
+    # It should be an array of type `string` in the format `<QUERY_PARAMETER_NAME>=<PARAMETER_VALUE>`
+    # **Note:** For more information about query parameters, refer here: 
+    # https://docs.microsoft.com/en-us/graph/query-parameters
+    # + return - a record `Event` if success. Else `error`.
```
Comment (abeykoon) [inline calendar/client.bal:61]: Guideline says 

When you have more than one returned items:
# + return - Generated string token,  an `auth:Error` occurred while generating the token, or else () if nothing is to be returned
# + return - `true` if authentication is successful, `false` otherwise, or else an `auth:Error` occurred while authenticating the credentials.

File: calendar/client.bal [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+    # + return - a stream of `Calendar` if success. Else `error`.
+    @display {label: "List Calendars"}
+    remote isolated function listCalendars(@display {label: "Optional Query Parameters"} string? queryParams = ()) 
+                                            returns stream<Calendar, error>|error {
+        string path = check createUrl([LOGGED_IN_USER, CALENDARS], queryParams);
+        CalendarStream objectInstance = check new (self.config, self.httpClient, path, queryParams);
+        stream<Calendar, error> finalStream = new (objectInstance);
+        return finalStream;
+    }
+}
```
Comment (abeykoon) [inline calendar/client.bal:261]: Code looks clean. Great!

File: calendar/Package.md (hunk 15 above)
Comment (Chuhaa) [reply calendar/Package.md:6]: In the package.md, we don't provide it manually as central has the API document link in the page.

File: calendar/tests/test.bal [hunk 19]
```diff
@@ -0,0 +1,421 @@
...
+    io:println("\n\n");
+}
+
+@test:Config {
+    enable: true,
+    groups: ["events"],
+    before: testCreateEvent,
+    after: testDeleteEvent
+}
+function testGetEventWithQueryParameters() {
```
Comment (abeykoon) [inline calendar/tests/test.bal:88]: It will be worth to describe what we test here in one sentence - wDYT?

File: build.gradle (hunk 11 above)
Comment (SkNuwanTissera) [reply build.gradle:87]: Changed in all places. 
/Thanks

File: calendar/Package.md (hunk 12 above)
Comment (SkNuwanTissera) [reply calendar/Package.md]: Yeah.I will change the introduction. Shall we change the Documentation guideline?
image

Comment (SkNuwanTissera) [reply calendar/Package.md]: changed the Intro. Added a comment in the doc. https://docs.google.com/document/d/1WZg1ix9MpKDTQ_OOm3NVqhQ5L0VGNaUQ1LRmjp-SMO4/edit?disco=AAAANWCuYwk

File: calendar/data_mappings.bal (hunk 13 above)
Comment (SkNuwanTissera) [reply calendar/data_mappings.bal:22]: Removed the data_mapping.bal 
/Thanks

File: calendar/stream_implementer.bal (hunk 14 above)
Comment (SkNuwanTissera) [reply calendar/stream_implementer.bal]: fixed in all places 
/Thanks

File: calendar/Package.md (hunk 15 above)
Comment (SkNuwanTissera) [reply calendar/Package.md:6]: Yeah. Let's go with that approach then.

File: calendar/client.bal (hunk 16 above)
Comment (SkNuwanTissera) [reply calendar/client.bal:34]: Changed.
/Thanks

File: calendar/client.bal (hunk 17 above)
Comment (SkNuwanTissera) [reply calendar/client.bal]: Changed in all the places.
/Thanks

File: calendar/client.bal (hunk 18 above)
Comment (SkNuwanTissera) [reply calendar/client.bal:261]: :-D. Thank you very much

File: calendar/tests/test.bal (hunk 19 above)
Comment (SkNuwanTissera) [reply calendar/tests/test.bal:88]: Yeah. It is worth discribing. Lets do

Comment (SkNuwanTissera) [reply calendar/tests/test.bal:92]: Added in all test cases
/Thanks

File: README.md [hunk 20]
```diff
@@ -1,2 +1,51 @@
...
+   * [Oracle](https://www.oracle.com/java/technologies/javase-jdk11-downloads.html)
+
+   * [OpenJDK](https://adoptopenjdk.net/)
+
+        > **Note:** Set the JAVA_HOME environment variable to the path name of the directory into which you installed JDK.
+
+2. Download and install [Ballerina Beta2](https://ballerina.io/). 
+
+### Building the source
+Execute the commands below to build from the source after installing Ballerina SL Beta 2 version.
```
Comment (shaniR) [inline README.md:30]: We don't need to mention this right? Since it's a prerequisite step?

File: calendar/Module.md [hunk 21]
```diff
@@ -0,0 +1,94 @@
+## Overview
+Ballerina connector for Microsoft Outlook Calendar is connecting the Calendar API in Microsoft Graph via Ballerina language easily.
```
Comment (shaniR) [inline calendar/Module.md:2]: Shall we remove unnecessary spaces and rephrase this as: "Ballerina connector for Microsoft Outlook Calendar is connecting the Calendar API in Microsoft Graph via Ballerina language easily. It provides the capability to perform CRUD (Create, Read, Update, and Delete) operations on  [Event](https://docs.microsoft.com/en-us/graph/api/resources/event?view=graph-rest-1.0) & [Calendar](https://docs.microsoft.com/en-us/graph/api/resources/event?view=graph-rest-1.0)  stored in Microsoft OneDrive. 	 
  
This module supports the Microsoft Outlook Calendar API v1.0 version."

File: calendar/Module.md [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+It provides capability to perform perform CRUD (Create, Read, Update, and Delete) operations on 
+[Event](https://docs.microsoft.com/en-us/graph/api/resources/event?view=graph-rest-1.0) &
+[Calendar](https://docs.microsoft.com/en-us/graph/api/resources/event?view=graph-rest-1.0) 
+stored in Microsoft OneDrive. 
+ 
+This module supports Microsoft Outlook Calendar API v1.0 version.
+ 
+### Prerequisites
+Before using this connector in your Ballerina application, complete the following:
+* Create [Microsoft Outlook Account](https://outlook.live.com/owa/)
```
Comment (shaniR) [inline calendar/Module.md:12]: Create a

File: calendar/Module.md [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+* Obtaining tokens
```
Comment (shaniR) [inline calendar/Module.md:13]: Obtain tokens

File: calendar/Module.md [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+    Follow [this link](https://docs.microsoft.com/en-us/graph/auth-v2-user#authentication-and-authorization-steps) and obtain the client ID, client secret and refresh token.
```
Comment (shaniR) [inline calendar/Module.md:14]: client ID, client secret, and refresh token.

File: calendar/Module.md [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+ 
+## Quickstart
+To use the outlook calendar connector in your Ballerina application, update the .bal file as follows:
+
+### Step 1: Import MS Outlook Calendar Package 
+import the ballerinax/microsoft.outlook.calendar module into the Ballerina project.
+```ballerina
```
Comment (shaniR) [inline calendar/Module.md:21]: Leave a line

File: calendar/Module.md [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+import ballerinax/microsoft.outlook.calendar;
+```
+### Step 2: Configure the connection to an existing Azure AD app
+You can now make the connection configuration using the OAuth2 refresh token grant config.
+```ballerina
```
Comment (shaniR) [inline calendar/Module.md:26]: Add an empty new line

File: calendar/Module.md [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+    }
+};
+calendar:Client calendarClient = check new(configuration);
+```
+### Step 3: Invoke  connector operation
+
+1. Now you can use the operations available within the connector. Note that they are in the form of remote operations.
+Following is an example on how to create a calendar using the connector.
+
+Create a calendar
```
Comment (shaniR) [inline calendar/Module.md:42]: Let's either use the title or the description.

File: calendar/Module.md [hunk 28]
```diff
@@ -0,0 +1,94 @@
...
+    log:printInfo(generatedEvent.toString());
+}
+   
+```
+2. Use `bal run` command to compile and run the Ballerina program.
+
+## Quick reference 
+
+* Get event details
+Ex
```
Comment (shaniR) [inline calendar/Module.md:84]: Remove the work Ex from all instances.

File: calendar/Package.md [hunk 29]
```diff
@@ -0,0 +1,20 @@
+Connects to Microsoft Outlook Calendar from Ballerina
+
+### Package overview
+The `outlook.calendar` is a [Ballerina](https://ballerina.io/) connector for Outlook Calendar.
```
Comment (shaniR) [inline calendar/Package.md:4]: Let's use the connector name. 
Microsoft Outlook Calendar

File: calendar/Package.md [hunk 30]
```diff
@@ -0,0 +1,20 @@
...
+This package provides the capability to easily create appointments & events, organize meetings, manage user's calendar etc.
```
Comment (shaniR) [inline calendar/Package.md:5]: create appointments, create events, organize meetings, manage a user's calendar, etc.

File: calendar/Package.md [hunk 31]
```diff
@@ -0,0 +1,20 @@
...
+
+#### Compatibility
+|                                 | Version                         |
+|---------------------------------|---------------------------------|
+| Ballerina Language              | Ballerina Swan Lake Beta2       |
+| Microsoft Outlook Calendar API  | v1.0                            |
+
+
+### Report issues
```
Comment (shaniR) [inline calendar/Package.md:14]: Use level ## heading

File: calendar/Package.md [hunk 32]
```diff
@@ -0,0 +1,20 @@
...
+To report bugs, request new features, start new discussions, view project boards, etc., go to the [Ballerina Microsoft calendar repository](https://github.com/ballerina-platform/module-ballerinax-microsoft.outlook.calendar)
+
+### Useful links
```
Comment (shaniR) [inline calendar/Package.md:17]: Use ## heading

**Review by shaniR — CHANGES_REQUESTED** [review]

File: README.md [hunk 33]
```diff
@@ -1,2 +1,51 @@
-# module-ballerinax-microsoft.outlook.calendar
-Connects to Microsoft Calendar using Ballerina.
+Ballerina Microsoft Outlook Connector
+======================================
+
+[![Build Status](https://github.com/ballerina-platform/module-ballerinax-microsoft.outlook.calendar/workflows/CI/badge.svg)](https://github.com/ballerina-platform/module-ballerinax-microsoft.outlook.calendar/actions?query=workflow%3ACI)
+[![GitHub Last Commit](https://img.shields.io/github/last-commit/ballerina-platform/module-ballerinax-microsoft.outlook.calendar.svg)](https://github.com/ballerina-platform/module-ballerinax-microsoft.outlook.calendar/commits/master)
+[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
+
+The `microsoft.outlook.calendar` is a [Ballerina](https://ballerina.io/) connector for Microsoft Outlook Calendar.
```
[the remark this answers is not in the document]
Comment (SkNuwanTissera) [reply README.md:8]: Sure. Let's change that way

File: README.md [hunk 34]
```diff
@@ -1,2 +1,51 @@
...
+
+[Microsoft Outlook Calendar](https://www.microsoft.com/en-ww/microsoft-365/outlook.calendar) is composed of the calendar and scheduling component of Outlook developed by Microsoft.
```
[the remark this answers is not in the document]
Comment (SkNuwanTissera) [reply README.md:10]: Removed unnessary lines

File: README.md [hunk 35]
```diff
@@ -1,2 +1,51 @@
...
+## Building from the source
+### Setting up the prerequisites
+
+1. Download and install Java SE Development Kit (JDK) version 11 (from one of the following locations).
+
+   * [Oracle](https://www.oracle.com/java/technologies/javase-jdk11-downloads.html)
+
+   * [OpenJDK](https://adoptopenjdk.net/)
+
+        > **Note:** Set the JAVA_HOME environment variable to the path name of the directory into which you installed JDK.
```
[the remark this answers is not in the document]
Comment (SkNuwanTissera) [reply README.md:25]: Added as in the agreed doc 
/Thanks

File: README.md (hunk 20 above)
Comment (SkNuwanTissera) [reply README.md:30]: Yeah. We will remove

File: calendar/Module.md (hunk 21 above)
Comment (SkNuwanTissera) [reply calendar/Module.md:2]: Rephrased as above /Thanks

File: calendar/Module.md (hunk 22 above)
Comment (SkNuwanTissera) [reply calendar/Module.md:12]: Corrected. /Thanks

File: calendar/Module.md (hunk 23 above)
Comment (SkNuwanTissera) [reply calendar/Module.md:13]: Corrected. /Thanks

File: calendar/Module.md (hunk 24 above)
Comment (SkNuwanTissera) [reply calendar/Module.md:14]: Shall we remove the trailing part? `and obtain the client ID, client secret and refresh token.`

Comment (SkNuwanTissera) [reply calendar/Module.md:14]: Corrected /Thanks

File: calendar/Module.md (hunk 26 above)
Comment (SkNuwanTissera) [reply calendar/Module.md:26]: Added empty line 
/Thanks

File: calendar/Module.md (hunk 27 above)
Comment (SkNuwanTissera) [reply calendar/Module.md:42]: Let's remove the title since the description explains the operation below.
/Thanks

File: calendar/Module.md (hunk 28 above)
Comment (SkNuwanTissera) [reply calendar/Module.md:84]: Removed `Ex` from all places

File: calendar/Package.md (hunk 30 above)
Comment (SkNuwanTissera) [reply calendar/Package.md:5]: Corrected as above /Thanks

File: calendar/Module.md (hunk 25 above)
Comment (SkNuwanTissera) [reply calendar/Module.md:21]: Left a line. /Thanks

File: calendar/Package.md (hunk 29 above)
Comment (SkNuwanTissera) [reply calendar/Package.md:4]: Changed. /Thanks

File: calendar/Package.md (hunk 31 above)
Comment (SkNuwanTissera) [reply calendar/Package.md:14]: Changed /Thanks

File: calendar/Package.md (hunk 32 above)
Comment (SkNuwanTissera) [reply calendar/Package.md:17]: Changed /Thanks

Comment (SkNuwanTissera) [reply calendar/Package.md:17]: Corrected in Module. md as well.
/Thanks

**Review by abeykoon — APPROVED** [review]

**Review by indikasampath2000 — APPROVED** [review]

File: calendar/client.bal [hunk 36]
```diff
@@ -0,0 +1,267 @@
...
+@display {
+    label: "Microsoft Calendar Client",
+    iconPath: "logo.svg"
+}
+public isolated client class Client {
+    private final http:Client httpClient;
+    final readonly & Configuration config;
+
+    # Initializes the connector. 
+    # During initialization you can pass either http:BearerTokenConfig
```
Comment (shaniR) [inline calendar/client.bal:34]: To initialize, you can either pass the http:BearerTokenConfig if you have a bearer token or the http:OAuth2RefreshTokenGrantConfig if you have OAuth 2.0 tokens.

File: calendar/types.bal [hunk 37]
```diff
@@ -0,0 +1,666 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+
+# Represents configuration parameters to create Microsoft Calendar.
+#
+# + clientConfig - OAuth client configuration
```
Comment (shaniR) [inline calendar/types.bal:21]: OAuth 2.0

File: calendar/client.bal (hunk 36 above)
Comment (SkNuwanTissera) [reply calendar/client.bal:34]: Hi Shani, 
This Client configuration record has been unified across connectors in the new release [1]. 
This record type accepts several authentication modes [2].

[1] https://github.com/ballerina-platform/module-ballerinax-microsoft.outlook.calendar/blob/4e2c2ad018816ffe0a3fee9c205f988472f2afe0/calendar/client.bal#L31
[2] https://github.com/ballerina-platform/module-ballerinax-microsoft.outlook.calendar/blob/4e2c2ad018816ffe0a3fee9c205f988472f2afe0/calendar/types.bal#L21

Thanks

File: calendar/types.bal (hunk 37 above)
Comment (SkNuwanTissera) [reply calendar/types.bal:21]: Hi Shani, 

This has been updated with new changes in record type [1].
Please refer to the new record documentation here [2].

[1] https://github.com/ballerina-platform/module-ballerinax-microsoft.outlook.calendar/blob/4e2c2ad018816ffe0a3fee9c205f988472f2afe0/calendar/client.bal#L31

[2] https://github.com/ballerina-platform/module-ballerinax-microsoft.outlook.calendar/blob/4e2c2ad018816ffe0a3fee9c205f988472f2afe0/calendar/types.bal#L19

Thanks
